### ISSUE TITLE
Fails to compile on CentOS 5, glibc-2.5, kernel 2.6.18-408.el5

### ISSUE THREAD

**Issue by thresheek** [acassen/keepalived#320]
The error is as follows:
```
gcc -I../include -I../../lib -O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m32 -march=i386 -mtune=generic -fasynchronous-unwind-tables -DETHERTYPE_IPV6=0x86dd  -I/usr/src/linux/include -I/usr/src/linux/include -Wall -Wunused -Wstrict-prototypes -D_KRNL_2_6_ -D_WITHOUT_LVS_ -D_WITHOUT_IPVS_SYNCD_ -D_WITH_VRRP_ -D_WITHOUT_VRRP_VMAC_ -D_WITHOUT_ADDR_GEN_MODE_ -D_WITHOUT_SNMP_ -D_WITHOUT_SNMP_KEEPALIVED_ -D_WITHOUT_SNMP_CHECKER_ -D_WITHOUT_SNMP_RFC_ -D_WITHOUT_SNMP_RFCV2_ -D_WITHOUT_SNMP_RFCV3_ -D_WITHOUT_LIBNL_ -D_WITH_VRRP_AUTH_ -D_WITHOUT_SO_MARK_ -D_WITHOUT_LIBIPTC_ -D_WITHOUT_LIBIPSET_ -D_WITHOUT_IPV4_DEVCONF_ -D_WITHOUT_IF_H_LINK_H_COLLISION_  -c smtp.c
smtp.c: In function 'smtp_connect':
smtp.c:584: error: 'SOCK_CLOEXEC' undeclared (first use in this function)
smtp.c:584: error: (Each undeclared identifier is reported only once
smtp.c:584: error: for each function it appears in.)
make[2]: *** [smtp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/home/thresh/rpmbuild/BUILD/keepalived-1.2.20/keepalived/core'
make[1]: *** [all] Error 1
make[1]: Leaving directory `/home/thresh/rpmbuild/BUILD/keepalived-1.2.20/keepalived'
make: *** [all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.13669 (%build)
```

**Comment by thresheek** [comment]
When hacking around that particular error with:
```
#ifndef SOCK_CLOEXEC
#define SOCK_CLOEXEC 0
#endif

#ifndef SOCK_NONBLOCK
#define SOCK_NONBLOCK 0
#endif
```

It fails further down the code on:
```
gcc -I../include -I../../lib -g -O2 -DETHERTYPE_IPV6=0x86dd  -I/usr/src/linux/include -I/usr/src/linux/include -Wall -Wunused -Wstrict-prototypes -D_KRNL_2_6_ -D_WITHOUT_LVS_ -D_WITHOUT_IPVS_SYNCD_ -D_WITH_VRRP_ -D_WITHOUT_VRRP_VMAC_ -D_WITHOUT_ADDR_GEN_MODE_ -D_WITHOUT_SNMP_ -D_WITHOUT_SNMP_KEEPALIVED_ -D_WITHOUT_SNMP_CHECKER_ -D_WITHOUT_SNMP_RFC_ -D_WITHOUT_SNMP_RFCV2_ -D_WITHOUT_SNMP_RFCV3_ -D_WITHOUT_LIBNL_ -D_WITH_VRRP_AUTH_ -D_WITHOUT_SO_MARK_ -D_WITHOUT_LIBIPTC_ -D_WITHOUT_LIBIPSET_ -D_WITHOUT_IPV4_DEVCONF_ -D_WITHOUT_IF_H_LINK_H_COLLISION_  -c vrrp_netlink.c
vrrp_netlink.c: In function ‘netlink_if_link_populate’:
vrrp_netlink.c:606: error: ‘IFLA_INFO_MAX’ undeclared (first use in this function)
vrrp_netlink.c:606: error: (Each undeclared identifier is reported only once
vrrp_netlink.c:606: error: for each function it appears in.)
vrrp_netlink.c:607: error: ‘IFLA_MACVLAN_MAX’ undeclared (first use in this function)
vrrp_netlink.c:640: error: ‘IFLA_LINKINFO’ undeclared (first use in this function)
vrrp_netlink.c:651: error: ‘IFLA_INFO_KIND’ undeclared (first use in this function)
vrrp_netlink.c:654: error: ‘IFLA_INFO_DATA’ undeclared (first use in this function)
vrrp_netlink.c:657: error: ‘IFLA_MACVLAN_MODE’ undeclared (first use in this function)
vrrp_netlink.c:658: error: ‘MACVLAN_MODE_PRIVATE’ undeclared (first use in this function)
vrrp_netlink.c:607: warning: unused variable ‘linkattr’
vrrp_netlink.c:606: warning: unused variable ‘linkinfo’
make[2]: *** [vrrp_netlink.o] Error 1
make[2]: Leaving directory `/home/thresh/keep/keepalived-1.2.20/keepalived/vrrp'
make[1]: *** [all] Error 1
make[1]: Leaving directory `/home/thresh/keep/keepalived-1.2.20/keepalived'
make: *** [all] Error 2
```

**Comment by pqarmitage** [comment]
The compile errors in vrrp_netlink.c are resolved in commit e7d9047 .

I'm looking to see what can be done regarding SOCK_CLOEXEC and SOCK_NONBLOCK

**Comment by thresheek** [comment]
Thank you for your work, @pqarmitage.

This indeed fixes the error in vrrp_netlink.c, and now it fails on vrrp_iprule.c:

```gcc -I../include -I../../lib -g -O2 -DETHERTYPE_IPV6=0x86dd  -I/usr/src/linux/include -I/usr/src/linux/include -Wall -Wunused -Wstrict-prototypes -D_KRNL_2_6_ -D_WITHOUT_LVS_ -D_WITHOUT_IPVS_SYNCD_ -D_WITH_VRRP_ -D_WITHOUT_VRRP_VMAC_ -D_WITHOUT_ADDR_GEN_MODE_ -D_WITHOUT_SNMP_ -D_WITHOUT_SNMP_KEEPALIVED_ -D_WITHOUT_SNMP_CHECKER_ -D_WITHOUT_SNMP_RFC_ -D_WITHOUT_SNMP_RFCV2_ -D_WITHOUT_SNMP_RFCV3_ -D_WITHOUT_LIBNL_ -D_WITH_VRRP_AUTH_ -D_WITHOUT_SO_MARK_ -D_WITHOUT_LIBIPTC_ -D_WITHOUT_LIBIPSET_ -D_WITHOUT_IPV4_DEVCONF_ -D_WITHOUT_IF_H_LINK_H_COLLISION_  -c vrrp_iprule.c
vrrp_iprule.c:25:29: error: linux/fib_rules.h: No such file or directory
vrrp_iprule.c: In function ‘netlink_rule’:
vrrp_iprule.c:77: error: ‘FRA_TABLE’ undeclared (first use in this function)
vrrp_iprule.c:77: error: (Each undeclared identifier is reported only once
vrrp_iprule.c:77: error: for each function it appears in.)
vrrp_iprule.c:91: error: ‘FRA_SRC’ undeclared (first use in this function)
vrrp_iprule.c:94: error: ‘FRA_DST’ undeclared (first use in this function)
make[2]: *** [vrrp_iprule.o] Error 1
make[2]: Leaving directory `/home/thresh/keep/keepalived-1.2.20/keepalived/vrrp'
make[1]: *** [all] Error 1
make[1]: Leaving directory `/home/thresh/keep/keepalived-1.2.20/keepalived'
make: *** [all] Error 2
```

**Comment by pqarmitage** [comment]
@thresheek. Rather than me submitting more commits that only partially resolve the problems you are having, I think it best if we work on another branch until everything is resolved.

If you could use branch Linux-2.6.18 from http://github.com/pqarmitage/keepalived, then once we have got everything resolved there, I can submit all the patches to @acassen.

On the above branch, I have included fixes for the SOCK_CLOEXEC and SOCK_NONBLOCK problem, and also the missing defines for FRA_TABLE, FRA_SRC, FRA_DST (these latter problems are caused because it was 2.6.19 that added in FIB routing support).

**Comment by thresheek** [comment]
Thank you @pqarmitage. I was able to compile keepalived using the Linux-2.6.18 branch (v1.2.2-890-ga4d6789) with the following configure options:

```
./configure    --disable-lvs --disable-lvs-syncd --disable-fwmark --disable-snmp --disable-sha1
```

When building with ```./configure``` w/o any arguments, it bails out with ```configure: error: No SO_MARK declaration in headers```, which is expected, I guess.

When building with ```./configure --disable-fwmark```, configure passes, but the build fails on:
```
gcc -I../include -I../../lib -g -O2 -DETHERTYPE_IPV6=0x86dd  -I/usr/src/linux/include -I/usr/src/linux/include -Wall -Wunused -Wstrict-prototypes -D_KRNL_2_6_ -D_WITH_LVS_ -D_HAVE_IPVS_SYNCD_ -D_WITH_VRRP_ -D_WITHOUT_VRRP_VMAC_ -D_WITHOUT_ADDR_GEN_MODE_ -D_WITHOUT_SNMP_ -D_WITHOUT_SNMP_KEEPALIVED_ -D_WITHOUT_SNMP_CHECKER_ -D_WITHOUT_SNMP_RFC_ -D_WITHOUT_SNMP_RFCV2_ -D_WITHOUT_SNMP_RFCV3_ -D_WITHOUT_LIBNL_ -D_WITH_VRRP_AUTH_ -D_WITHOUT_SO_MARK_ -D_WITHOUT_LIBIPTC_ -D_WITHOUT_LIBIPSET_ -D_WITHOUT_IPV4_DEVCONF_ -D_WITHOUT_IF_H_LINK_H_COLLISION_ -D_WITHOUT_SOCK_NONBLOCK_ -D_WITHOUT_SOCK_CLOEXEC_ -D_WITHOUT_FIB_ROUTING_ -D_HAVE_LINUX_NET_IF_H_COLLISION_  -c ipvswrapper.c
ipvswrapper.c: In function ‘get_modprobe’:
ipvswrapper.c:1106: error: ‘O_CLOEXEC’ undeclared (first use in this function)
ipvswrapper.c:1106: error: (Each undeclared identifier is reported only once
ipvswrapper.c:1106: error: for each function it appears in.)
make[2]: *** [ipvswrapper.o] Error 1
make[2]: Leaving directory `/home/thresh/keepalived/keepalived/check'
make[1]: *** [all] Error 1
make[1]: Leaving directory `/home/thresh/keepalived/keepalived'
make: *** [all] Error 2
```

I must also say that I was unable to open the issue on your repo, so that's why I replied here.

Thanks again.

**Comment by pqarmitage** [comment]
@thresheek, many thanks for testing and your feedback. I think updating the issue report he is the best way to provide updates.

--disable-snmp and --disable-sha1 are the default options, and so shouldn't need to be specified. Please let me know omitting them doesn't work for you.

I have pushed a couple of further commits to my tree, so that it should now do the following:
- configure will no longer complain if SO_MARK is not defined, but just act as though --disable-fwmark had been specified
- checks if O_CLOEXEC is not defined, and works around it

The ideal would be if we can get keepalived to build without having to specify any options to configure, and it just sorts out what is needed (as it is supposed to do).

If you could pull my latest commits and try doing a clean build running configure with no options and let me know if it works or there are still problems, then if it works I'll push the changes upstream.

**Comment by thresheek** [comment]
Using v1.2.2-893-g7a939fd, keepalived builds fine, without specifying any --disable-* arguments.

I've also checked with and without --disable-snmp --disable-sha1 (as well as --disable-lvs --disable-lvs-syncd --disable-fwmark) and did not encounter any issues.

I've also checked v1.2.2-893-g7a939fd against CentOS7 and Ubuntu 16.04, also no issues.

Those commits fix all the issues I had, thank you very much!
